SERVER_EXTENSIONS.md

```diff
@@ -0,0 +1,274 @@
+# Server extensions
+
+> The Ruby LSP server extensions system is currently experimental and subject to changes in the API
+
+## Motivation and goals
+
+Having editor features that are specific to certain tools can be incredibly helpful for programmers. For example, Rails
+related features or specific formatters. Typically, language servers are aimed at providing features for a particular
+programming language (like Ruby!) and not specific frameworks or other tools. This is reasonable since not every Ruby
+programmer will use the same set of tools.
+
+Trying to include tool specific functionality directly in the Ruby LSP would not scale well given the large number of
+tools in the Ruby ecosystem. It would also create a bottleneck for tool authors since all changes would need to be
+merged into the Ruby LSP's main codebase.
+
+Creating language servers for each tool is also not an adequate solution (e.g.: Rails LSP, Sinatra LSP). While this
+might sometimes be easier for the maintainers, it places the burden on the user to figure out which language servers to
+install to get a complete set of features and how to configure them so that there are no conflicts.
+
+For these reasons, the Ruby LSP ships with a server extension system that gem authors can use to enhance the behavior of
+the base language server with tool specific functionality. The goals of the extension system are
+
+- To allow gem authors to export Ruby LSP extensions from their own gems
+- To allow LSP features to be enhanced by extensions that are present in the application the developer is currently
+working on, without requiring extra configuration from the user
+- To seamlessly integrate with the base features of the Ruby LSP
+
+## Guidelines
```

Comment:
I personally find the difference between automatic and user-initiated requests important but I wasn't aware of it until reading this. So perhaps adding it as the second point too?

> It's generally annoying and diverts attention from the current task

It may be worth mentioning that we learned this through experiences, not just an opinion.

Reply:
I added another point related to automatic and user initiated requests.

Regarding the last point, I honestly think it's fine to keep it as is - even if it was an opinion.